### ISSUE TITLE
Whoops page, possibly SQL error? Failed to start MariaDB?

### ISSUE THREAD

**Issue by damnjuhl** [firefly-iii/firefly-iii#2917]
I am running Firefly III version 4.8.2

**Description**
I've been using Firefly the past 15 days or so, today I noticed my 'budget' and 'net worth' were not being calculated. Shortly after I restarted my Pi, and returned to the page saying 'Whoops, check the logs' I went to firefly-iii/storage/logs and selected the one with todays date, `ff3-audit-2019-12-26.log` this is all that I see, I'm not sure how to reproduce this error or what exactly is even going wrong.

```
[2019-12-26 04:07:33] local.INFO: User visits homepage.  {"path":"GET:http://173.27.98.55/public","IP":"173.27.98.55","user":"damnjuhl@gmail.com"}
[2019-12-26 04:39:15] local.INFO: User visits homepage.  {"path":"GET:http://173.27.98.55/public","IP":"173.27.98.55","user":"damnjuhl@gmail.com"}
[2019-12-26 04:40:02] local.INFO: Store new transaction over API. {"group_title":"","error_if_duplicate_hash":false,"transactions":[{"type":"deposit","date":"2019-12-25 00:00:00","order":null,"currency_id":null,"currency_code":null,"for$
[2019-12-26 04:40:03] local.INFO: User visits homepage.  {"path":"GET:http://173.27.98.55/public","IP":"173.27.98.55","user":"damnjuhl@gmail.com"}
[2019-12-26 04:40:16] local.INFO: User visits homepage.  {"path":"GET:http://173.27.98.55/public","IP":"173.27.98.55","user":"damnjuhl@gmail.com"}
[2019-12-26 05:31:12] local.INFO: User visits homepage.  {"path":"GET:http://173.27.98.55/public","IP":"173.27.98.55","user":"damnjuhl@gmail.com"}
[2019-12-26 08:12:25] local.INFO: User visits homepage.  {"path":"GET:http://192.168.0.69/public","IP":"192.168.0.107","user":"damnjuhl@gmail.com"}
[2019-12-26 09:36:43] local.INFO: User visits homepage.  {"path":"GET:http://192.168.0.69/public","IP":"192.168.0.107","user":"damnjuhl@gmail.com"}
[2019-12-26 18:40:21] local.INFO: User visits homepage.  {"path":"GET:http://173.27.98.55/public/home","IP":"63.152.13.187","user":"damnjuhl@gmail.com"}
[2019-12-26 18:40:51] local.INFO: Store new transaction over API. {"group_title":"","error_if_duplicate_hash":false,"transactions":[{"type":"transfer","date":"2019-12-26 00:00:00","order":null,"currency_id":null,"currency_code":null,"fo$
[2019-12-26 18:40:52] local.INFO: User visits homepage.  {"path":"GET:http://173.27.98.55/public/home","IP":"63.152.13.187","user":"damnjuhl@gmail.com"}
[2019-12-26 18:41:04] local.INFO: User visits homepage.  {"path":"GET:http://173.27.98.55/public","IP":"63.152.13.187","user":"damnjuhl@gmail.com"}
[2019-12-26 18:42:26] local.INFO: Store new transaction over API. {"group_title":"","error_if_duplicate_hash":false,"transactions":[{"type":"transfer","date":"2019-12-26 00:00:00","order":null,"currency_id":null,"currency_code":null,"fo$
[2019-12-26 18:42:27] local.INFO: User visits homepage.  {"path":"GET:http://173.27.98.55/public","IP":"63.152.13.187","user":"damnjuhl@gmail.com"}
[2019-12-26 18:42:52] local.INFO: User visits homepage.  {"path":"GET:http://173.27.98.55/public","IP":"63.152.13.187","user":"damnjuhl@gmail.com"}
[2019-12-26 23:51:37] local.INFO: User visits homepage.  {"path":"GET:http://192.168.0.69/public","IP":"192.168.0.107","user":"damnjuhl@gmail.com"}
[2019-12-26 23:51:59] local.INFO: User visits homepage.  {"path":"GET:http://192.168.0.69/public","IP":"192.168.0.107","user":"damnjuhl@gmail.com"}
[2019-12-26 23:52:17] local.INFO: User visits homepage.  {"path":"GET:http://192.168.0.69/public","IP":"192.168.0.107","user":"damnjuhl@gmail.com"}
[2019-12-26 23:52:21] local.INFO: User visits homepage.  {"path":"GET:http://192.168.0.69/public","IP":"192.168.0.107","user":"damnjuhl@gmail.com"}

```
![image](https://user-images.githubusercontent.com/15189394/71493966-26dc2600-2809-11ea-8715-90c1aa66486e.png)

When I navigated to the page via my phone a different output can be read at this screen shot:

![image](https://user-images.githubusercontent.com/15189394/71493973-47a47b80-2809-11ea-8dee-3d51b0c894f4.png)

This makes me believe it's an SQL error but restarting it doesn't do it. I attempted to start mysql to ensure it was running, 

```
/etc/init.d/mysql start
[....] Starting mysql (via systemctl): mysql.service==== AUTHENTICATING FOR org.freedesktop.systemd1.manage-units ===
Authentication is required to start 'mariadb.service'.
Multiple identities can be used for authentication:
 1.  ,,, (pi)
 2.  root
Choose identity to authenticate as (1-2): 1
Password:
==== AUTHENTICATION COMPLETE ===
Job for mariadb.service failed because a fatal signal was delivered to the control process.
See "systemctl status mariadb.service" and "journalctl -xe" for details.
 failed!

```

I then input, `systemctl status mariadb.service` and received this output: 

```
● mariadb.service - MariaDB 10.3.17 database server
   Loaded: loaded (/lib/systemd/system/mariadb.service; enabled; vendor preset: enabled)
   Active: activating (start) since Thu 2019-12-26 18:09:08 CST; 511ms ago
     Docs: man:mysqld(8)
           https://mariadb.com/kb/en/library/systemd/
  Process: 24195 ExecStartPre=/usr/bin/install -m 755 -o mysql -g root -d /var/run/mysqld (code=exited, status=0/SUCCESS)
  Process: 24196 ExecStartPre=/bin/sh -c systemctl unset-environment _WSREP_START_POSITION (code=exited, status=0/SUCCESS)
  Process: 24198 ExecStartPre=/bin/sh -c [ ! -e /usr/bin/galera_recovery ] && VAR= ||   VAR=`/usr/bin/galera_recovery`; [ $? -eq 0 ]   && systemctl set-environment _WSREP_START_POSITION=$VAR || exit 1 (code=exited, status=0/SUCCESS)
 Main PID: 24246 ((mysqld))
    Tasks: 1 (limit: 2200)
   Memory: 1.1M
   CGroup: /system.slice/mariadb.service
           └─24246 (mysqld)

Dec 26 18:09:08 danpi systemd[1]: Starting MariaDB 10.3.17 database server...

```
Here is `journalctl -xe`

```
--
-- The process' exit code is 'killed' and its exit status is 6.
Dec 26 18:09:16 danpi systemd[1]: mariadb.service: Failed with result 'signal'.
-- Subject: Unit failed
-- Defined-By: systemd
-- Support: https://www.debian.org/support
--
-- The unit mariadb.service has entered the 'failed' state with result 'signal'.
Dec 26 18:09:16 danpi systemd[1]: Failed to start MariaDB 10.3.17 database server.
-- Subject: A start job for unit mariadb.service has failed
-- Defined-By: systemd
-- Support: https://www.debian.org/support
--
-- A start job for unit mariadb.service has finished with a failure.
--
-- The job identifier is 22322 and the job result is failed.
Dec 26 18:09:21 danpi systemd[1]: mariadb.service: Service RestartSec=5s expired, scheduling restart.
Dec 26 18:09:21 danpi systemd[1]: mariadb.service: Scheduled restart job, restart counter is at 359.
-- Subject: Automatic restarting of a unit has been scheduled
-- Defined-By: systemd
-- Support: https://www.debian.org/support
--
-- Automatic restarting of the unit mariadb.service has been scheduled, as the result for
-- the configured Restart= setting for the unit.
Dec 26 18:09:21 danpi systemd[1]: Stopped MariaDB 10.3.17 database server.
-- Subject: A stop job for unit mariadb.service has finished
-- Defined-By: systemd
-- Support: https://www.debian.org/support
--
-- A stop job for unit mariadb.service has finished.
--
-- The job identifier is 22382 and the job result is done.
Dec 26 18:09:21 danpi systemd[1]: Starting MariaDB 10.3.17 database server...
-- Subject: A start job for unit mariadb.service has begun execution
-- Defined-By: systemd
-- Support: https://www.debian.org/support
--
-- A start job for unit mariadb.service has begun execution.
--
-- The job identifier is 22382.
Dec 26 18:09:22 danpi mysqld[24380]: 2019-12-26 18:09:22 0 [Note] /usr/sbin/mysqld (mysqld 10.3.17-MariaDB-0+deb10u1) starting as process 24380 ...
Dec 26 18:09:23 danpi systemd[1]: mariadb.service: Main process exited, code=killed, status=6/ABRT
-- Subject: Unit process exited
-- Defined-By: systemd
-- Support: https://www.debian.org/support
--
-- An ExecStart= process belonging to unit mariadb.service has exited.
--
-- The process' exit code is 'killed' and its exit status is 6.
Dec 26 18:09:23 danpi systemd[1]: mariadb.service: Failed with result 'signal'.
-- Subject: Unit failed
-- Defined-By: systemd
-- Support: https://www.debian.org/support
--
-- The unit mariadb.service has entered the 'failed' state with result 'signal'.
Dec 26 18:09:23 danpi systemd[1]: Failed to start MariaDB 10.3.17 database server.
-- Subject: A start job for unit mariadb.service has failed
-- Defined-By: systemd
-- Support: https://www.debian.org/support
--
-- A start job for unit mariadb.service has finished with a failure.
--
-- The job identifier is 22382 and the job result is failed.
lines 5730-5792/5792 (END)

```

This is where I'm stuck at, I'm not sure why it won't start/why anything changed in the first place/how to fix it. Any help, pointers, insight or knowledge would be wonderful. Hopefully this is a simple fix, it had been working flawlessly for two weeks.

**Comment by JC5** [comment]
There should be another file with today’s date. That one contains the technical stuff. 

From these errors I really can’t tell why the database fails to start unfortunately.

**Comment by damnjuhl** [comment]
![image](https://user-images.githubusercontent.com/15189394/71533765-4d10cd00-28c0-11ea-960a-b06328a0bdb4.png)

I posted the most recent log that I could find, is there a different place I should be looking for a more current log file?

Found a log file for mysql, this was most recent output I could find:

```
2019-12-26 16:52:58 0 [Note] /usr/sbin/mysqld (initiated by: unknown): Normal shutdown
2019-12-26 16:52:58 0 [Note] Event Scheduler: Purging the queue. 0 events
2019-12-26 16:52:58 0 [Note] InnoDB: FTS optimize thread exiting.
2019-12-26 16:52:58 0 [Note] InnoDB: Starting shutdown...
2019-12-26 16:52:58 0 [Note] InnoDB: Dumping buffer pool(s) to /var/lib/mysql/ib_buffer_pool
2019-12-26 16:52:58 0 [Note] InnoDB: Buffer pool(s) dump completed at 191226 16:52:58
2019-12-26 16:52:59 0 [Note] InnoDB: Shutdown completed; log sequence number 7423307; transaction id 8777
2019-12-26 16:52:59 0 [Note] InnoDB: Removed temporary tablespace data file: "ibtmp1"
2019-12-26 16:52:59 0 [Note] /usr/sbin/mysqld: Shutdown complete

2019-12-26 16:53:11 0 [Note] InnoDB: Using Linux native AIO
2019-12-26 16:53:11 0 [Note] InnoDB: Mutexes and rw_locks use GCC atomic builtins
2019-12-26 16:53:11 0 [Note] InnoDB: Uses event mutexes
2019-12-26 16:53:11 0 [Note] InnoDB: Compressed tables use zlib 1.2.11
2019-12-26 16:53:11 0 [Note] InnoDB: Number of pools: 1
2019-12-26 16:53:11 0 [Note] InnoDB: Using generic crc32 instructions
2019-12-26 16:53:11 0 [Note] InnoDB: Initializing buffer pool, total size = 128M, instances = 1, chunk size = 128M
2019-12-26 16:53:11 0 [Note] InnoDB: Completed initialization of buffer pool
2019-12-26 16:53:11 0 [Note] InnoDB: If the mysqld execution user is authorized, page cleaner thread priority can be changed. See the man page of setpriority().
2019-12-26 16:53:12 0 [ERROR] InnoDB: Space id and page no stored in the page, read in are [page id: space=4294967295, page number=4294967295], should be [page id: space=0, page number=405]
2019-12-26 16:53:12 0 [ERROR] InnoDB: Database page corruption on disk or a failed file read of tablespace innodb_system page [page id: space=0, page number=405]. You may have to recover from a backup.
2019-12-26 16:53:12 0 [Note] InnoDB: Page dump in ascii and hex (16384 bytes):
 len 16384; hex ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff$
InnoDB: End of page dump
2019-12-26 16:53:13 0 [Note] InnoDB: Uncompressed page, stored checksum in field1 4294967295, calculated checksums for field1: crc32 612169467, innodb 913626112,  page type 65535 == PAGE TYPE CORRUPTED.none 3735928559, stored checksum i$
2019-12-26 16:53:13 0 [Note] InnoDB: It is also possible that your operating system has corrupted its own file cache and rebooting your computer removes the error. If the corrupt page is an index page. You can also try to fix the corrup$
2019-12-26 16:53:13 0 [ERROR] [FATAL] InnoDB: Aborting because of a corrupt database page.
191226 16:53:13 [ERROR] mysqld got signal 6 ;
This could be because you hit a bug. It is also possible that this binary
or one of the libraries it was linked against is corrupt, improperly built,
or misconfigured. This error can also be caused by malfunctioning hardware.

To report this bug, see https://mariadb.com/kb/en/reporting-bugs

We will try our best to scrape up some info that will hopefully help
diagnose the problem, but since we have already crashed,
something is definitely wrong and this may fail.

Server version: 10.3.17-MariaDB-0+deb10u1
key_buffer_size=134217728
read_buffer_size=131072
max_used_connections=0
max_threads=153
thread_count=0
It is possible that mysqld could use up to
key_buffer_size + (read_buffer_size + sort_buffer_size)*max_threads = 466218 K  bytes of memory
Hope that's ok; if not, decrease some variables in the equation.

Thread pointer: 0x0
Attempting backtrace. You can use the following information to find out
where mysqld died. If you see no messages after this, something went
terribly wrong...
stack_bottom = 0x0 thread_stack 0x49000
The manual page at http://dev.mysql.com/doc/mysql/en/crashing.html contains
information that should help you find out what is causing the crash.
Writing a core file...
Working directory at /var/lib/mysql
Resource Limits:
Limit                     Soft Limit           Hard Limit           Units
                                                                                                     [ File 'error.log' is unwritable ]
```

**Comment by JC5** [comment]
There should be files with names like `ff3-cli-2019-12-28.log` or `ff3-fpm-fcgi-2019-12-27.log`.

But it appears the problem is in your database:

```
[ERROR] InnoDB: Space id and page no stored in the page, read in are [page id: space=4294967295, page number=4294967295], should be [page id: space=0, page number=405]
[ERROR] InnoDB: Database page corruption on disk or a failed file read of tablespace innodb_system page [page id: space=0, page number=405]. You may have to recover from a backup.
[Note] InnoDB: It is also possible that your operating system has corrupted its own file cache and rebooting your computer removes the error. If the corrupt page is an index page. You can also try to fix the corrup$
[ERROR] [FATAL] InnoDB: Aborting because of a corrupt database page.
191226 16:53:13 [ERROR] mysqld got signal 6 ;
```

**Comment by damnjuhl** [comment]
Idk what is going on, I get everything working, I can access it for days, a week, I get my main accounts setup and input a few days back worth of transactions and everything seems okay. All the numbers add up, haven't been kicked any errors, no issues. Then I add budgets and bills and expense accounts and my credit card and everything looks to be working. Once everything was working I safely shut down and tried to back up the SD card drive to my computer, after that went with no errors I plug it back in and get whoops error after a restart. I've tried asking where the logs are but only see audit logs which tell me hey you added all these categories and such. And didn't seem like the logs I got asked for last time this happened, wasn't told where those would be just that they'd be there so idk what to do besides be sad
